install_script.md

```diff
@@ -0,0 +1,154 @@
+    "after": ["com.thirdcorp.thirdlibrary:thirdlibrary:1.2"]
+}
+```
+
+`before` and `after` allow this project's classpath library to be prioritized when multiple classpath libraries are present. If a library mentioned in `before` or `after` is not installed, this does not cause it to be installed - the entry does nothing. In `before` and `after` entries, the group ID and artifact ID are mandatory, but the version is optional (if not present then it matches all versions). Versions are an exact match.
+
+`name` must contain exactly a group ID, artifact ID, and version number.
+
+MODIP and non-MODIP (e.g. Mojang) libraries are all considered together for sorting. i.e. you may request to install a library before one Mojang library and after another.
+
+This example library is only required on OSX. On other platforms, this install step is a no-op.
+The launcher MAY still download the OSX natives and add the library entry, with the OSX-only rule, in case the instance is later moved to a computer running OSX.
+
+## `setMainClass` step
+
+Updates the main class. Mainly used by loaders and standalone mods.
+
+Example:
+
+```
+{
+    "type": "setMainClass",
+    "from": ["net.minecraft.client.main.Main"],
+    "to": "net.fabricmc.loader.launch.knot.KnotClient"
+}
+```
+
+`from` specifies what the main class should have been before. If multiple actions of this type are executed (from multiple mods), they must be linked in a chain, starting from the vanilla's main class, and then the last step in the chain indicates the actual main class. If they can't be linked in a chain, the installed mods are not compatible. `from` must be a list. The previous main class must match any value in the list.
+
+## Client/server fork step
+
+Uses a different installation method on the client than on the server.
+
+Example:
+
+```
+{
+    "type": "client_server_fork",
+    "client_steps": [
+    ],
+    "server_steps": [
+    ]
+}
+```
+
+Note: `client_steps` and `server_steps` are always lists, and must be present even if empty.
+
+## Framework fork step
+
+Allows a mod to support multiple mod loaders.
+
+```
+{
+    "type": "framework_fork",
+    "framework_steps": {
+        "<framework id>": [
+        ],
+        "<other framework id>": [
+        ]
+    },
+    "preferred": "<framework id>"
+}
+```
+
+Note: values in `framework_steps` are always lists, and must be present even if empty.
+If no entry is present for the current framework, the mod is incompatible with it.
+If the current framework is unknown (i.e. no mods are installed yet), the launcher should install the one in `preferred` (which must be one of the options).
+If *multiple* supported frameworks are present (i.e. Patchwork), the launcher MAY choose a preferred one, ask the user, or give up (fall back to manual installation).
+
+If this mechanism is used to support multiple frameworks, neither framework should be listed as a dependency, as that would require it to be installed!
```

Comment:
Good question. What do you think about putting dependencies inside forks?

Reply:
I think it could work. But I don't see much benefit of doing that instead of putting them in the file.

Reply:
The benefit is, obviously, that you can have different dependencies on different forks.

Reply:
I mean that you put the dependency in the file’s dependency list. You could have Fabric API listed as a dependency for the `mod name-fabric.jar` file, and have nothing for the `modname-forge.jar` file. This is already supported by the spec. The only potential issue is mods that work for both Fabric and Forge.

Reply:
If an installation might need multiple files, how exactly does that work? Does it merge together the dependency lists for each file that the installation script references?

Reply:
Merging them together was the idea, yeah.

Reply:
What if it has no files, or the same file with different dependencies?

Reply:
You have a point. I haven't really thought about no files, so I think dependencies in forks is a good idea.

Reply:
If forks can have dependencies, then maybe "framework forks" don't need to exist. You could have a fork based on nothing at all, where one side has Fabric as a dependency, and the other side has Forge as a dependency. And the loader would try both sides to see which one works, and if they both work, it would have to ask the user whether to install the Fabric or Forge configuration.

Does that seem sensible?

Reply:
That seems sensible to me.